### ISSUE TITLE
`model().res([])` creates an empty residual branch during graph build

### ISSUE THREAD

**Issue by nm-z** [nm-z/recipe-dev#825]
Observed failing phase: `training`.

Diagnostic: `thread 'main' ... panic at recipe.rs:16118:13: residual branch must contain an operation`.

Public entrypoint used in report: `recipe.data("data/numeric/sample_subfolders").target("target")`, `let model = recipe.model().layer(16).res([]).layer(1).loss(mse);`, then `recipe.train().optimizer(adamw).lr(0.001).seed(17).epochs(1).log(all).fp(32).save(&bundle).run(&model, &data)`.

Evidence:
- `Model::res<const N: usize>(&self, parts: [Block; N]) -> Self` pushes `Operation::Residual(parts.into())` with no guard for `N == 0`, so `.res([])` is accepted.
- `compile` lowers each top-level block, dispatching residual blocks through `lower_residual`.
- `lower_residual` first checks `require(!parts.is_empty(), "residual branch must contain an operation")?;`.
- The panic occurs before the reported training loop can progress to inference, matching `observed=phase:training` and no `inference` phase log.

Hypothesis:
- The earliest observed operation causing failure is the public `res([])` call in the model DSL, which creates an empty residual branch that is only rejected during compile-time lowering in training.

## Machine provenance

- Discoverer: Recipe composition traversal
- Failure fingerprint: `e8b26bc1c97df09c`
- Recipe source commit: `b384815c89c3db25b7d7587cf3e0c75de7b51aa9`
- Embedded Recipe `recipe.rs` blob: `35cd2fffcc44db98b053168b9eebb2194d4d8999`
- Source identity: The packet captures the exact working-tree Recipe source independently of the Recipe Machine commit.
- Classifier provider: `codex`
- Classifier model: `gpt-5.3-codex-spark`
- Classifier effort: `xhigh`

### Classifier decision

```json
{"verdict":"new","title":"`model().res([])` creates an empty residual branch during graph build","body":"Observed failing phase: `training`.\n\nDiagnostic: `thread 'main' ... panic at recipe.rs:16118:13: residual branch must contain an operation`.\n\nPublic entrypoint used in report: `recipe.data(\"data/numeric/sample_subfolders\").target(\"target\")`, `let model = recipe.model().layer(16).res([]).layer(1).loss(mse);`, then `recipe.train().optimizer(adamw).lr(0.001).seed(17).epochs(1).log(all).fp(32).save(&bundle).run(&model, &data)`.\n\nEvidence:\n- `Model::res<const N: usize>(&self, parts: [Block; N]) -> Self` pushes `Operation::Residual(parts.into())` with no guard for `N == 0`, so `.res([])` is accepted.\n- `compile` lowers each top-level block, dispatching residual blocks through `lower_residual`.\n- `lower_residual` first checks `require(!parts.is_empty(), \"residual branch must contain an operation\")?;`.\n- The panic occurs before the reported training loop can progress to inference, matching `observed=phase:training` and no `inference` phase log.\n\nHypothesis:\n- The earliest observed operation causing failure is the public `res([])` call in the model DSL, which creates an empty residual branch that is only rejected during compile-time lowering in training.","rationale":"The evidence directly ties a single user-observable operation (`.res([])`) to the exact guard that triggers the panic. This is a structural validation gap in model construction: invalid zero-length residuals are not blocked at API entry but are rejected during training compile. No fix is proposed, only the root cause and observed failure linkage are reported."}
```

## Deterministic failure packet

id=e8b26bc1c97df09c
base=commit=b384815c89c3db25b7d7587cf3e0c75de7b51aa9 recipe_source=35cd2fffcc44db98b053168b9eebb2194d4d8999
cursor=cursor:0 next:1 composition:0
data=path:data/numeric/sample_subfolders
configuration=cursor:0 seed:17
expected=the generated public Recipe composition trains and infers with finite values
observed=phase:training message:note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
output=phase:training
thread 'main' (1546682) panicked at recipe.rs:16118:13:
residual branch must contain an operation
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
replay=phase:training message:note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace stable:true
command=RECIPE_FORCE_CPU=1 target/debug/recipe test.rs
reproduction:
```rust
use recipe::*;
use std::path::{Path, PathBuf};

fn phase(value: &str) {
    if let Some(path) = std::env::var_os("RECIPE_COMPOSITION_PHASE_PATH") {
        std::fs::write(path, value).expect("cannot write phase");
    }
}

fn width(path: &Path) -> usize {
    let text = std::fs::read_to_string(path).expect("cannot read saved model");
    let mut shape = text.lines().find_map(|line| line.strip_prefix("shape ")).expect("saved model has no shape").split_whitespace();
    shape.next().expect("saved model has no channels").parse::<usize>().expect("saved channels are invalid") * shape.next().expect("saved model has no length").parse::<usize>().expect("saved length is invalid")
}

fn main() {
    let root = std::env::var_os("RECIPE_TRIAL_DIRECTORY").map(PathBuf::from).unwrap_or_else(std::env::temp_dir);
    let bundle = root.join(format!("recipe-model-{}.ogdl", std::process::id()));
    phase("setup");
    let data = recipe.data("data/numeric/sample_subfolders").target("target");
    let model = recipe.model().layer(16).res([]).layer(1).loss(mse);
    phase("training");
    let report = recipe.train().optimizer(adamw).lr(0.001).seed(17).epochs(1).log(all).fp(32).save(&bundle).run(&model, &data);
    assert!(report.final_loss().is_finite());
    phase("inference");
    let output = recipe.infer(&bundle, &vec![0.0; width(&bundle)]);
    assert!(!output.is_empty() && output.iter().all(|value| value.is_finite()));
    std::fs::remove_file(bundle).expect("cannot remove saved model");
}
```
backend=cpu

**Comment by nm-z** [comment]
The observed error is caused by an undefined value in the LLVM IR during training, which is likely due to an empty residual branch created by `model().res([])`. This matches issue #825, which discusses the creation of an empty residual branch during graph build that can lead to undefined behavior.

## Machine provenance

- Discoverer: Recipe composition traversal
- Failure fingerprint: `27965a1776d570db`
- Recipe source commit: `ec32b36432407f70fc77c0c0cae56339ccab5203`
- Embedded Recipe `recipe.rs` blob: `e149eb47c95b7992c006a1a6a8f18616b5e07a85`
- Source identity: The packet captures the exact working-tree Recipe source independently of the Recipe Machine commit.
- Classifier provider: `ollama`
- Classifier model: `qwen3:8b`
- Classifier effort: `default`

### Classifier decision

```json
{
  "verdict": "comment",
  "issue": 825,
  "title": "model().res([]) creates an empty residual branch during graph build",
  "body": "The observed error is caused by an undefined value in the LLVM IR during training, which is likely due to an empty residual branch created by `model().res([])`. This matches issue #825, which discusses the creation of an empty residual branch during graph build that can lead to undefined behavior.",
  "rationale": "The error message indicates a use of an undefined value in the LLVM IR, which is a direct consequence of an improperly constructed residual branch. Issue #825 describes a similar problem where an empty residual branch leads to undefined behavior during training. The observed failure aligns with the confirmed cause in that issue, making it the strongest match."
}
```

## Deterministic failure packet

id=27965a1776d570db
base=commit=ec32b36432407f70fc77c0c0cae56339ccab5203 recipe_source=e149eb47c95b7992c006a1a6a8f18616b5e07a85
cursor=cursor:15642 next:15643 composition:15642
data=path:data/numeric/sample_subfolders
configuration=cursor:15642 seed:17
expected=the generated public Recipe composition trains and infers with finite values
observed=phase:training message:NVIDIA LLVM IR compiler failed: .recipe-native--.ll:3937:12: error: use of undefined value '%n2.normalize.stats.mean.stored'
output=phase:training
measured archy:nv4 5.503831e7 work/s 0.000028963s/dispatch to-host 151.6 MB/s 37µs from-host 206.8 MB/s 1ms
route archy:nv4 rows 64 predicted epoch 5.454134542s = computation 5.452944114 + transfers 0.000042259 + synchronization 0.001123785 + persistent-state 0.000024383
selected route archy:nv4 predicted epoch 5.454134542s

thread 'main' (26462) panicked at recipe.rs:21854:13:
NVIDIA LLVM IR compiler failed: /home/nate/.cache/recipe/native/recipe-native-9192ed305c2bf8ce/.recipe-native-26462-0.ll:3937:12: error: use of undefined value '%n2.normalize.stats.mean.stored'
 3937 | store half %n2.normalize.stats.mean.stored, ptr addrspace(1) %n2.normalize.stats.mean.context.ptr, align 1
      |            ^
1 error generated.
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
replay=phase:training message:NVIDIA LLVM IR compiler failed: .recipe-native--.ll:3937:12: error: use of undefined value '%n2.normalize.stats.mean.stored' stable:true
command=RECIPE_DEVICE=nv4 target/debug/recipe test.rs
reproduction:
```rust
use recipe::*;
use std::path::{Path, PathBuf};

fn phase(value: &str) {
    if let Some(path) = std::env::var_os("RECIPE_COMPOSITION_PHASE_PATH") {
        std::fs::write(path, value).expect("cannot write phase");
    }
}

fn width(path: &Path) -> usize {
    let text = std::fs::read_to_string(path).expect("cannot read saved model");
	let mut shape = text.lines().find_map(|line| line.trim().strip_prefix("shape ")).expect("saved model has no shape").split_whitespace();
    shape.next().expect("saved model has no channels").parse::<usize>().expect("saved channels are invalid") * shape.next().expect("saved model has no length").parse::<usize>().expect("saved length is invalid")
}

fn main() {
    let root = std::env::var_os("RECIPE_TRIAL_DIRECTORY").map(PathBuf::from).unwrap_or_else(std::env::temp_dir);
    let bundle = root.join(format!("recipe-model-{}.ogdl", std::process::id()));
    phase("setup");
    let data = recipe.data("data/numeric/sample_subfolders").target("target");
    let model = recipe.model().perc(8).selu().norm(layer).res([layer(8).silu().norm(rms), rnn(8).relu().qi(4).nf, attn(1).width(32).prelu().norm(l2).qi(4).nf]).scale(2.0).layer(1).loss(mae);
    phase("training");
    let report = recipe.train().optimizer(adamw).lr(0.001).seed(17).epochs(20).log(all).fp(16).save(&bundle).run(&model, &data);
    assert!(report.final_loss().is_finite());
    phase("inference");
    let output = recipe.infer(&bundle, &vec![0.0; width(&bundle)]);
    assert!(!output.is_empty() && output.iter().all(|value| value.is_finite()));
    std::fs::remove_file(bundle).expect("cannot remove saved model");
}
```
backend=nv4
